core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -0,0 +1,643 @@
+ * to call the function returning the HW capacity.
+ */
+static uint8_t caam_hash_limit;
+
+/*
+ * Returns the reference to the driver context
+ *
+ * @ctx  API Context
+ */
+static struct crypto_hash *to_hash_ctx(struct crypto_hash_ctx *ctx)
+{
+	assert(ctx && ctx->ops == &hash_ops);
+
+	return container_of(ctx, struct crypto_hash, hash_ctx);
+}
+
+/*
+ * Free the internal hashing data context
+ *
+ * @ctx    [in/out] Caller context variable
+ */
+static void do_free_intern(struct hashdata *ctx)
+{
+	HASH_TRACE("Free Context (0x%" PRIxPTR ")", (uintptr_t)ctx);
+
+	if (ctx) {
+		/* Free the descriptor */
+		caam_free_desc(&ctx->descriptor);
+
+		/* Free the Temporary buffer */
+		caam_free_buf(&ctx->blockbuf.buf);
+
+		/* Free the context register */
+		caam_free_buf(&ctx->ctx);
+	}
+}
+
+/*
+ * Allocate the internal hashing data context
+ *
+ * @ctx    [in/out] Caller context variable
+ */
+static enum CAAM_Status do_allocate_intern(struct hashdata *ctx)
+{
+	TEE_Result ret = CAAM_OUT_MEMORY;
+
+	HASH_TRACE("Allocate Context (0x%" PRIxPTR ")", (uintptr_t)ctx);
+
+	/* Allocate the descriptor */
+	ctx->descriptor = caam_alloc_desc(MAX_DESC_ENTRIES);
+	if (!ctx->descriptor) {
+		HASH_TRACE("Allocation descriptor error");
+		goto exit_alloc;
+	}
+
+	/* Initialize the block buffer */
+	ctx->blockbuf.filled = 0;
+	ctx->blockbuf.max = ctx->alg->size_block;
+
+	/* Allocate the CAAM Context register */
+	ret = caam_alloc_align_buf(&ctx->ctx, ctx->alg->size_ctx);
+
+#ifdef HASH_DEBUG
+	if (ret != CAAM_NO_ERROR)
+		HASH_TRACE("Allocation context register error");
+#endif
+
+	/* Ensure buffer length is 0 */
+	ctx->ctx.length = 0;
+
+exit_alloc:
+	if (ret != CAAM_NO_ERROR) {
+		/* Free all data allocated */
+		do_free_intern(ctx);
+	}
+
+	return ret;
+}
+
+/*
+ * Free the SW hashing data context
+ *
+ * @ctx    [in/out] Caller context variable
+ */
+static void do_free(struct crypto_hash_ctx *ctx)
+{
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	HASH_TRACE("Free Context (0x%" PRIxPTR ")", (uintptr_t)hash->ctx);
+
+	if (hash->ctx) {
+		do_free_intern(hash->ctx);
+		caam_free(hash->ctx);
+	}
+
+	free(hash);
+}
+
+/*
+ * Initialization of the Hash operation
+ *
+ * @ctx   Operation Software context
+ */
+static TEE_Result do_init(struct crypto_hash_ctx *ctx)
+{
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+	struct hashdata *hashdata = hash->ctx;
+
+	HASH_TRACE("Hash Init (0x%" PRIxPTR ")", (uintptr_t)hashdata);
+	if (!hashdata)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Initialize the block buffer */
+	hashdata->blockbuf.filled = 0;
+
+	/* Ensure Context length is 0 */
+	hashdata->ctx.length = 0;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Update the Hash operation
+ *
+ * @ctx   Operation Software context
+ * @data  Data to hash
+ * @len   Data length
+ */
+static TEE_Result do_update(struct crypto_hash_ctx *ctx, const uint8_t *data,
+			    size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum CAAM_Status retstatus = CAAM_FAILURE;
+
```

Comment:
Please drop the empty lines here in the variable declarations.

Reply:
ok